### DISCUSSION TITLE
Hashing user passwords in backend server

### DISCUSSION THREAD

**Discussion by eeroleppalehto** [eeroleppalehto/Jahtirekisteri#195] [Ideas]
# Hashing

For user management to be possible, users need a username and password so that they can be authenticated. These usernames and passwords need to be stored in the database but passwords should not be stored as plaintext as that is a major security threat. As solution for this the passwords need to hashed before storing to database. But the question is which method of hashing should the server use?

Here are two popular libraries that provide methods for hashing in **Nodejs**:

### [bcrypt](https://www.npmjs.com/package/bcrypt?activeTab=readme)

Provides a compute heavy hashing algorithm to makes guessing passwords hard as each guess takes a lot of time. Includes support for salting passwords.

### [crypto](https://nodejs.org/api/crypto.html#determining-if-crypto-support-is-unavailable)

A built-in package of Nodejs that provides `pbkdf2Sync()` method for hashing. Includes support for salting passwords.

--------

## Feedback

Share your thoughts on these method. What issues might arise with either of the presented solutions and if you have any other solutions on your mind please share them.

**Comment by AleksanteriLohja** [comment]
**bcrypt**

**Pros:**

Enhanced Security: bcrypt uses a compute-intensive algorithm, making it difficult to guess passwords since each attempt is time-consuming. This reduces the feasibility of brute-force attacks.
Built-in Salting: It automatically adds a salt to the passwords, ensuring that the hash is unique for each password, even if the same password is used multiple times.
Widely Trusted: As a well-established method, bcrypt has a strong track record and community support, underlining its reliability.

**Cons:**

Performance Overhead: The algorithm's complexity can be a drawback on servers with high traffic or limited resources.
Scalability Issues: As user authentication requests increase, bcrypt's time-consuming nature might become a bottleneck.

**crypto (pbkdf2Sync)**

**Pros:**

Built-in Node.js Package: Being a part of Node.js, it simplifies implementation as no additional libraries are required.
Customizability: It allows adjusting parameters like the number of iterations, balancing between security and performance.
Supports Salting: This is crucial for preventing attacks like rainbow table attacks.

**Cons:**

Complex Configuration: Properly setting up crypto (like specifying iterations) can be more challenging for less experienced developers.
Comparatively Less Robust: While pbkdf2Sync is secure, it may not be as resilient against certain types of attacks as bcrypt.

**Summary:**

The choice between bcrypt and crypto's pbkdf2Sync method in Node.js depends on the application's specific needs and environment. If security is the primary concern and some performance trade-off is acceptable, bcrypt is a strong choice. Conversely, if customization and performance are critical, crypto's pbkdf2Sync might be preferable.
